src/sdk/pynni/nni/bohb_advisor/bohb_advisor.py

```diff
@@ -0,0 +1,594 @@
+                                    max_budget=self.max_budget, optimize_mode=self.optimize_mode)
+        next_n, next_r = self.brackets[self.curr_s].get_n_r()
+        logger.debug(
+            'new SuccessiveHalving iteration, next_n=%d, next_r=%d', next_n, next_r)
+        # rewrite with TPE
+        generated_hyper_configs = self.brackets[self.curr_s].get_hyperparameter_configurations(
+            next_n, next_r, self.cg)
+        self.generated_hyper_configs = generated_hyper_configs.copy()
+
+    def handle_request_trial_jobs(self, data):
+        """recerive the number of request and generate trials
+
+        Parameters
+        ----------
+        data: int
+            number of trial jobs that nni manager ask to generate
+        """
+        # Receive new request
+        self.credit += data
+
+        for _ in range(self.credit):
+            self._request_one_trial_job()
+
+        return True
+
+    def _request_one_trial_job(self):
+        """get one trial job, i.e., one hyperparameter configuration.
+
+        Returns
+        -------
+        dict:
+            one hyperparameter configuration
+            0: 'parameter_id', id of new hyperparameter
+            1: 'parameter_source', 'algorithm'
+            2: 'parameters', value of new hyperparameter
+        """
+        if not self.generated_hyper_configs:
+            ret = {
+                'parameter_id': '-1_0_0',
+                'parameter_source': 'algorithm',
+                'parameters': ''
+            }
+            send(CommandType.NoMoreTrialJobs, json_tricks.dumps(ret))
+            return True
+        assert self.generated_hyper_configs
+        params = self.generated_hyper_configs.pop()
+        ret = {
+            'parameter_id': params[0],
+            'parameter_source': 'algorithm',
+            'parameters': params[1]
+        }
+        self.parameters[params[0]] = params[1]
+        send(CommandType.NewTrialJob, json_tricks.dumps(ret))
+        self.credit -= 1
+        return True
+
+    def handle_update_search_space(self, data):
+        """change json format to ConfigSpace format dict<dict> -> configspace
+
+        Parameters
+        ----------
+        data: JSON object
+            search space of this experiment
+
+        Returns
+        -------
+        ConfigSpace:
+            search space in ConfigSpace format
+        """
+        search_space = data
+        cs = CS.ConfigurationSpace()
+        for var in search_space:
+            _type = str(search_space[var]["_type"])
+            if _type == 'choice':
+                cs.add_hyperparameter(CSH.CategoricalHyperparameter(
+                    var, choices=search_space[var]["_value"]))
+            elif _type == 'randint':
+                cs.add_hyperparameter(CSH.UniformIntegerHyperparameter(
+                    var, lower=0, upper=search_space[var]["_value"][0]))
+            elif _type == 'uniform':
```

Comment:
can we merge some situation? such as "uniform " and "quniform"? (suggestion)

Reply:
Not clear about the suggestion? "uniform " is different from "quniform", "quniform" required "q" as an arg but uniform don't need. Even if it is merged, we still have to use an if-else logic to distinguish q=1 or q = other value...